File: examples/cpp/Beta/fastsam_parser.cpp
#include <algorithm>
#include <cstddef>
#include <iostream>
#include <memory>
#include <opencv2/opencv.hpp>
#include <string>

#include "depthai/beta/datatypes.hpp"
#include "depthai/beta/nodes.hpp"
#include "depthai/depthai.hpp"

int main() {
    dai::Pipeline pipeline;

    const std::string modelSlug = "luxonis/fastsam-s:512x288";
    dai::NNModelDescription modelDescription{modelSlug};
    modelDescription.platform = pipeline.getDefaultDevice()->getPlatformAsString();
    dai::NNArchive modelArchive(dai::getModelFromZoo(modelDescription));

    auto cameraNode = pipeline.create<dai::node::Camera>()->build();
    auto neuralNetwork = pipeline.create<dai::node::NeuralNetwork>()->build(cameraNode, modelArchive);
    auto parserNode = pipeline.create<dai::beta::node::FastSAMParser>()->build(neuralNetwork->out, modelArchive);

    auto frameQueue = neuralNetwork->passthrough.createOutputQueue();
    auto outputQueue = parserNode->out.createOutputQueue();
    auto configQueue = parserNode->inputConfig.createInputQueue();
    auto config = parserNode->initialConfig;

    pipeline.start();
    std::cout << "Controls: '+' increase confidence threshold, '-' decrease it, 'q' quit." << std::endl;

    while(pipeline.isRunning()) {
        auto frameMessage = frameQueue->get<dai::ImgFrame>();
        auto parserOutput = outputQueue->get<dai::SegmentationMask>();
        cv::Mat frame = frameMessage->getCvFrame();
        cv::Mat mask = parserOutput->getCvMask();
        cv::resize(mask, mask, frame.size(), 0.0, 0.0, cv::INTER_NEAREST);
        cv::Mat scaledMask;
        mask.convertTo(scaledMask, CV_8U, 37.0);
        cv::Mat coloredMask;
        cv::applyColorMap(scaledMask, coloredMask, cv::COLORMAP_TURBO);
        coloredMask.setTo(cv::Scalar(0, 0, 0), mask == 255);
        cv::addWeighted(frame, 0.6, coloredMask, 0.4, 0.0, frame);

        cv::imshow("FastSAMParser", frame);
        const int key = cv::waitKey(1);
        if(key == 'q') break;
        if(key == '+') {
            config->confidenceThreshold = std::min(1.0f, config->confidenceThreshold + 0.1f);
            configQueue->send(config);
            std::cout << "Confidence threshold: " << config->confidenceThreshold << std::endl;
        } else if(key == '-') {
            config->confidenceThreshold = std::max(0.0f, config->confidenceThreshold - 0.1f);
            configQueue->send(config);
            std::cout << "Confidence threshold: " << config->confidenceThreshold << std::endl;
        }
    }

    return 0;
}

File: examples/python/Beta/fastsam_parser.py
#!/usr/bin/env python3

import cv2
import depthai as dai
import numpy as np


def main() -> None:
    with dai.Pipeline() as pipeline:
        modelSlug = "luxonis/fastsam-s:512x288"
        modelDescription = dai.NNModelDescription(
            modelSlug,
            platform=pipeline.getDefaultDevice().getPlatformAsString(),
        )
        modelArchive = dai.NNArchive(dai.getModelFromZoo(modelDescription))

        cameraNode = pipeline.create(dai.node.Camera).build()
        neuralNetwork = pipeline.create(dai.node.NeuralNetwork).build(cameraNode, modelArchive)
        parserNode = pipeline.create(dai.beta.node.FastSAMParser).build(
            neuralNetwork.out,
            modelArchive,
        )

        frameQueue = neuralNetwork.passthrough.createOutputQueue()
        outputQueue = parserNode.out.createOutputQueue()
        configQueue = parserNode.inputConfig.createInputQueue()
        config = parserNode.initialConfig

        pipeline.start()
        print("Controls: '+' increase confidence threshold, '-' decrease it, 'q' quit.")

        while pipeline.isRunning():
            frameMessage = frameQueue.get()
            parserOutput = outputQueue.get()
            frame = frameMessage.getCvFrame()
            mask = np.asarray(parserOutput.getCvMask(), dtype=np.uint8)
            mask = cv2.resize(mask, (frame.shape[1], frame.shape[0]), interpolation=cv2.INTER_NEAREST)
            coloredMask = cv2.applyColorMap(mask * 37, cv2.COLORMAP_TURBO)
            coloredMask[mask == 255] = 0
            frame = cv2.addWeighted(frame, 0.6, coloredMask, 0.4, 0)


            cv2.imshow("FastSAMParser", frame)
            key = cv2.waitKey(1)
            if key == ord("q"):
                break
            if key == ord("+"):
                config.confidenceThreshold = min(1.0, config.confidenceThreshold + 0.1)
                configQueue.send(config)
                print(f"Confidence threshold: {config.confidenceThreshold:.1f}")
            elif key == ord("-"):
                config.confidenceThreshold = max(0.0, config.confidenceThreshold - 0.1)
                configQueue.send(config)
                print(f"Confidence threshold: {config.confidenceThreshold:.1f}")


if __name__ == "__main__":
    main()
